### PR TITLE
premisses to BindImageTexture fix

### PR DESCRIPTION
**Pull request by mp3butcher**
I followed your advice for getModifiedCount 
http://forum.openscenegraph.org/viewtopic.php?t=17102

### CODE REVIEW

**Comment by mp3butcher** [comment]
i'll reopen this later

File: include/osg/Texture [hunk 1]
```diff
@@ -771,6 +774,7 @@ class OSG_EXPORT Texture : public osg::StateAttribute
           * for all graphics contexts. */
         virtual void releaseGLObjects(State* state=0) const;
 
+        virtual unsigned int &getModifiedCount(unsigned int contextID) const = 0;
```
Comment (filnet) [inline include/osg/Texture]: Seems like your other changes make it possible to not return a reference (which contradicts the const nature of the method).

File: include/osg/Texture2DMultisample [hunk 2]
```diff
@@ -46,7 +46,12 @@ class OSG_EXPORT Texture2DMultisample : public Texture
         {
           return GL_TEXTURE_2D_MULTISAMPLE;
         }
-
```
Comment (filnet) [inline include/osg/Texture2DMultisample]: those empty lines should be kept

File: include/osg/Texture (hunk 1 above)
Comment (mp3butcher) [reply include/osg/Texture]: the method is const because it concerns a mutable attributes...

File: include/osg/Texture2DMultisample (hunk 2 above)
Comment (mp3butcher) [reply include/osg/Texture2DMultisample]: haven't see that..

**Comment by mp3butcher** [comment]
I close this until I achieve to make all examples work with it....I've got a strange bug in gpucull I haven't with others my branches...

**Comment by mp3butcher** [comment]
I've found the problem:
a strange behavior of stateset.getTextureAttribute(tu,type) : member searched is always set to 0 !?

File: include/osg/Texture (hunk 1 above)
Comment (filnet) [reply include/osg/Texture]: yes, but your changes make it possible to not return a reference to the mutable.

File: src/osg/TextureRectangle.cpp [hunk 3]
```diff
@@ -202,10 +202,10 @@ void TextureRectangle::apply(State& state) const
         {
             _subloadCallback->subload(*this, state);
         }
-        else if (_image.valid() && getModifiedCount(contextID) != _image->getModifiedCount())
+        else if (_image.valid() && _modifiedCount[contextID] != _image->getModifiedCount())
         {
             // update the modified count to show that it is up to date.
-            getModifiedCount(contextID) = _image->getModifiedCount();
+            _modifiedCount[contextID] = _image->getModifiedCount();
```
Comment (filnet) [inline src/osg/TextureRectangle.cpp]: this is one example where getModified() had to return a ref (and does not have to anymore)

**Comment by openscenegraph** [comment]
Really?  After everything I suggested about "what to" and what not do you go follow the path of "what not to do".

If you are modifying dozens of files to implement a niche StateAttribute there is something seriously wrong with the approach.  18 Files changed is a yet another sliurge over lots of the OSG.  The skill of a programmer is measured by the quality of their work not the quantity.

You need to think more of surgeon with a scalpe and not a butcher with a cleaver.  

These are real lessons to take away, to practice in your own code and especially submissions to the OSG.  You will end up a better programmer.

For this particular issue when I will be implement my previous suggestion.  I won't be able to get to it today, but hopefully midweek.

**Comment by mp3butcher** [comment]
Really! I just obey your guidelines...
I did modify a lot of useless calls to getModifiied in the textures because it surely became cost....

**Comment by openscenegraph** [comment]
I have checked in two commits, the first is the addition of a bool Texutre::isDirty(contextID) method and it's implementation in the various Texture subclasses:

https://github.com/openscenegraph/OpenSceneGraph/commit/b8e5f22cb7ad2b24451e55e8e117b9e4a404104f

The second commit is adding usage of the Texture::isDirty(contextID) into the BindImageTexture class, and usage of the State::applyTexturtueAttribute() API to work with the OSG's lazy state updating.  

https://github.com/openscenegraph/OpenSceneGraph/commit/68bb67312e417fc6d93bf8856609587114a380c6

The number of files changed was 8 headers and 1 cpp.  I haven't modified the examples, but this was always a separate commit and am happy for others to pitch in if required.  I haven't changed BindImageTexture to be a TextureAttribute as at this stage I think we can get away without doing this.

I would recommend studying these commits to see how one can efficiently and safely encapsulate functionality that is useful for multiple purposes and it's implementation is kept orthogonal to existing functionality.  All while changing significantly less files. The key principle is to Do more, risk less and do so with less code changes.

File: examples/osggpucull/osggpucull.cpp [hunk 4]
```diff
@@ -858,33 +853,31 @@ struct ResetTexturesCallback : public osg::StateSet::Callback
...
-                img->dirty();
-        }
-        for(it=texUnitsDirtyParams.begin(), eit=texUnitsDirtyParams.end(); it!=eit; ++it)
-        {
-            osg::TextureBuffer* tex = dynamic_cast<osg::TextureBuffer*>( stateset->getTextureAttribute(*it,osg::StateAttribute::TEXTURE) );
-            if(tex!=NULL)
-                tex->dirtyTextureParameters();
+    {
+           osg::BindImageTexture*imb = 0;
+           const osg::StateSet::AttributeList & texlist= stateset->getTextureAttributeList()[*it];
```
Comment (mp3butcher) [inline examples/osggpucull/osggpucull.cpp]: this dirty part is because ss.getTexAttribute(tu,type) fails : member searched is always set to 0 !?

**Comment by mp3butcher** [comment]
Okay!..., sorry,  I unconsciously put bad will on it.. A little more thinking would have surely lead me to this impl...
I would make a pr to fix examples but an other one is previously required to fix:
a strange behavior of stateset.getTextureAttribute(tu,type) : member searched is always set to 0 !?
Does simply add a 3rd (uint member) param with default value seams enough or do i have to add an other method ?

**Pull request closed without merging**